File: mojo/public/cpp/bindings/lib/pipe_control_message_handler.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/public/cpp/bindings/lib/pipe_control_message_handler.h"

#include "base/logging.h"
#include "mojo/public/cpp/bindings/lib/message_builder.h"
#include "mojo/public/cpp/bindings/lib/pipe_control_message_handler_delegate.h"
#include "mojo/public/cpp/bindings/lib/validation_util.h"
#include "mojo/public/interfaces/bindings/pipe_control_messages.mojom.h"

namespace mojo {
namespace internal {

PipeControlMessageHandler::PipeControlMessageHandler(
    PipeControlMessageHandlerDelegate* delegate)
    : delegate_(delegate) {}

PipeControlMessageHandler::~PipeControlMessageHandler() {}

// static
bool PipeControlMessageHandler::IsPipeControlMessage(const Message* message) {
  return !IsValidInterfaceId(message->interface_id());
}

bool PipeControlMessageHandler::Accept(Message* message) {
  if (!Validate(message))
    return false;

  if (message->name() == pipe_control::kRunOrClosePipeMessageId)
    return RunOrClosePipe(message);

  NOTREACHED();
  return false;
}

bool PipeControlMessageHandler::Validate(const Message* message) {
  if (message->name() == pipe_control::kRunOrClosePipeMessageId) {
    if (!ValidateMessageIsRequestWithoutResponse(message))
      return false;
    return ValidateMessagePayload<
        pipe_control::internal::RunOrClosePipeMessageParams_Data>(message);
  }

  return false;
}

bool PipeControlMessageHandler::RunOrClosePipe(Message* message) {
  pipe_control::internal::RunOrClosePipeMessageParams_Data* params =
      reinterpret_cast<
          pipe_control::internal::RunOrClosePipeMessageParams_Data*>(
          message->mutable_payload());
  params->DecodePointersAndHandles(message->mutable_handles());

  pipe_control::RunOrClosePipeMessageParamsPtr params_ptr;
  Deserialize_(params, &params_ptr, nullptr);

  if (params_ptr->input->is_peer_associated_endpoint_closed_event()) {
    return delegate_->OnPeerAssociatedEndpointClosed(
        params_ptr->input->get_peer_associated_endpoint_closed_event()->id);
  }
  if (params_ptr->input->is_associated_endpoint_closed_before_sent_event()) {
    return delegate_->OnAssociatedEndpointClosedBeforeSent(
        params_ptr->input->get_associated_endpoint_closed_before_sent_event()
            ->id);
  }

  DVLOG(1) << "Unsupported command in a RunOrClosePipe message pipe control "
           << "message. Closing the pipe.";
  return false;
}

}  // namespace internal
}  // namespace mojo

File: mojo/public/js/lib/pipe_control_message_handler.js
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

define("mojo/public/js/lib/pipe_control_message_handler", [
  "mojo/public/interfaces/bindings/pipe_control_messages.mojom",
  "mojo/public/js/codec",
  "mojo/public/js/interface_types",
  "mojo/public/js/validator",
], function(pipeControlMessages, codec, types, validator) {

  var Validator = validator.Validator;

  function validateControlRequestWithoutResponse(message) {
    var messageValidator = new Validator(message);
    var error = messageValidator.validateMessageIsRequestWithoutResponse();
    if (error != validator.validationError.NONE) {
      throw error;
    }

    if (message.getName() != pipeControlMessages.kRunOrClosePipeMessageId) {
      throw new Error("Control message name is not kRunOrClosePipeMessageId");
    }

    // Validate payload.
    error = pipeControlMessages.RunOrClosePipeMessageParams.validate(
        messageValidator, message.getHeaderNumBytes());
    if (error != validator.validationError.NONE) {
      throw error;
    }
  }

  function runOrClosePipe(message, delegate) {
    var reader = new codec.MessageReader(message);
    var runOrClosePipeMessageParams = reader.decodeStruct(
        pipeControlMessages.RunOrClosePipeMessageParams);
    var event = runOrClosePipeMessageParams.input
        .peer_associated_endpoint_closed_event;
    return delegate.onPeerAssociatedEndpointClosed(event.id,
        event.disconnect_reason);
  }

  function isPipeControlMessage(message) {
    return !types.isValidInterfaceId(message.getInterfaceId());
  }

  function PipeControlMessageHandler(delegate) {
    this.delegate_ = delegate;
  }

  PipeControlMessageHandler.prototype.accept = function(message) {
    validateControlRequestWithoutResponse(message);
    return runOrClosePipe(message, this.delegate_);
  };

  var exports = {};
  exports.PipeControlMessageHandler = PipeControlMessageHandler;
  exports.isPipeControlMessage = isPipeControlMessage;

  return exports;
});
